tls/s2n_prf.c

```diff
@@ -78,38 +79,131 @@ static int s2n_sslv3_prf(union s2n_prf_working_space *ws, struct s2n_blob *secre
     return 0;
 }
 
-static int s2n_p_hash(union s2n_prf_working_space *ws, s2n_hmac_algorithm alg, struct s2n_blob *secret,
+static int s2n_p_hash_init(union s2n_prf_working_space *ws, s2n_hmac_algorithm alg, const struct s2n_blob *secret)
+{
+    int r = 0;
+
+    eq_check(ws->tls.mac_key, NULL);
+    notnull_check(ws->tls.mac_key = EVP_PKEY_new_mac_key(EVP_PKEY_HMAC, NULL, secret->data, secret->size));
+
+    eq_check(ws->tls.md_ctx, NULL);
+#if S2N_OPENSSL_VERSION_AT_LEAST(1,1,0) && !defined(LIBRESSL_VERSION_NUMBER)
+    notnull_check(ws->tls.md_ctx = EVP_MD_CTX_new());
+#else
+    notnull_check(ws->tls.md_ctx = EVP_MD_CTX_create());
```

Comment:
If we're going to have a dynamic allocation for every connection, lets move it to `s2n_connection_new`